File: lap2d_green/lap2d_green.c
/*
 * lap2d_green.c -- C equivalent of lap2d_green.m
 *
 *   G(x,y) = -log(|x-y|^2) / (4*pi)
 *
 * Each rep k = 0..NREP-1 shifts target X coords by k*DT and adds the
 * resulting NT-by-NS kernel matrix into the accumulator `acc`.  The
 * final fingerprint is taken on `acc`, so every rep contributes to
 * the printed answer (no hoisting possible).
 *
 * Storage layout matches MATLAB column-major NT-by-NS:
 *   acc[j*NT + i]  for target i in [0,NT), source j in [0,NS).
 *
 * Build: see build.sh
 */

#define _POSIX_C_SOURCE 199309L

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <math.h>
#include <time.h>

#ifdef _OPENMP
#include <omp.h>
#endif

#ifndef NS
#define NS 4000
#endif
#ifndef NT
#define NT 4000
#endif
#ifndef NREP
#define NREP 5
#endif
#ifndef DT
#define DT 0.01
#endif

#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

#ifdef __FAST_MATH__
#define BUILD_LABEL "fast-math"
#else
#define BUILD_LABEL "strict"
#endif

static inline uint64_t xs64(uint64_t *s) {
    uint64_t x = *s;
    x ^= x << 13;
    x ^= x >> 7;
    x ^= x << 17;
    *s = x;
    return x;
}

static inline double xs64_double(uint64_t *s) {
    return (double)(xs64(s) >> 11) * (1.0 / 9007199254740992.0);
}

static double now_seconds(void) {
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return (double)ts.tv_sec + (double)ts.tv_nsec * 1e-9;
}

/* Add -log(|t-s|^2)/(4*pi) (with target X shifted by k*dt) into acc. */
static void accumulate_kernel(const double *src_x, const double *src_y,
                              const double *tgt_x, const double *tgt_y,
                              double *acc, int k, double dt) {
    const double fourpi = 4.0 * M_PI;
    const double dx = (double)k * dt;
    #pragma omp parallel for schedule(static)
    for (int j = 0; j < NS; j++) {
        double sx = src_x[j], sy = src_y[j];
        double *acc_col = acc + (size_t)j * NT;
        for (int i = 0; i < NT; i++) {
            double rx = (tgt_x[i] + dx) - sx;
            double ry = tgt_y[i] - sy;
            double r2 = rx*rx + ry*ry;
            acc_col[i] += -log(r2) / fourpi;
        }
    }
}

int main(void) {
    static double src_x[NS], src_y[NS];
    static double tgt_x[NT], tgt_y[NT];

    size_t total = (size_t)NS * (size_t)NT;
    double *acc = (double *)malloc(total * sizeof(double));
    if (!acc) { fprintf(stderr, "alloc failed\n"); return 1; }

    /* Inputs -- same xorshift64 seed and draw order as the .m file. */
    uint64_t state = 2463534242ULL;
    for (int j = 0; j < NS; j++) {
        src_x[j] = xs64_double(&state) * 2.0 - 1.0;
        src_y[j] = xs64_double(&state) * 2.0 - 1.0;
    }
    for (int i = 0; i < NT; i++) {
        tgt_x[i] = xs64_double(&state) * 2.0 - 1.0 + 3.0;
        tgt_y[i] = xs64_double(&state) * 2.0 - 1.0;
    }

    /* Warm up (one full pass through the kernel; result discarded). */
    memset(acc, 0, total * sizeof(double));
    accumulate_kernel(src_x, src_y, tgt_x, tgt_y, acc, 0, DT);

    /* Timed loop: accumulate NREP shifted kernel matrices into acc. */
    memset(acc, 0, total * sizeof(double));
    double t0 = now_seconds();
    for (int k = 0; k < NREP; k++) {
        accumulate_kernel(src_x, src_y, tgt_x, tgt_y, acc, k, DT);
    }
    double t1 = now_seconds();
    double elapsed = t1 - t0;

    /* Cross-language fingerprint. */
    int i_mid_m = NT / 2;
    int j_mid_m = NS / 2;
    int i_mid_c = i_mid_m - 1;
    int j_mid_c = j_mid_m - 1;

    double a_11  = acc[0];
    double a_mid = acc[(size_t)j_mid_c * NT + i_mid_c];
    double a_NN  = acc[(size_t)(NS-1) * NT + (NT-1)];

    double vmin = acc[0], vmax = acc[0], csum = 0.0;
    for (int j = 0; j < NS; j++) {
        const double *col = acc + (size_t)j * NT;
        for (int i = 0; i < NT; i++) {
            double v = col[i];
            csum += v;
            if (v < vmin) vmin = v;
            if (v > vmax) vmax = v;
        }
    }

    int nthreads = 1;
#ifdef _OPENMP
    #pragma omp parallel
    {
        #pragma omp single
        nthreads = omp_get_num_threads();
    }
#endif

    printf("=== C lap2d_green ===\n");
    printf("build          = %s\n", BUILD_LABEL);
    printf("threads        = %d\n", nthreads);
    printf("NS=%d  NT=%d  NREP=%d  DT=%.4f\n", NS, NT, NREP, (double)DT);
    printf("elapsed_total  = %.6f s\n", elapsed);
    printf("elapsed_per    = %.6f s\n", elapsed / NREP);
    printf("throughput     = %.3e pts/s\n",
           (double)NS * (double)NT * (double)NREP / elapsed);
    printf("acc(1,1)       = %.17e\n", a_11);
    printf("acc(%d,%d) = %.17e\n", i_mid_m, j_mid_m, a_mid);
    printf("acc(NT,NS)     = %.17e\n", a_NN);
    printf("min(acc)       = %.17e\n", vmin);
    printf("max(acc)       = %.17e\n", vmax);
    printf("sum(acc)       = %.17e\n", csum);

    free(acc);
    return 0;
}

File: lap2d_green/lap2d_green.m
function lap2d_green(nthreads)
% LAP2D_GREEN  Benchmark direct evaluation of the Laplace 2D Green's function.
%
%   G(x,y) = -log(|x-y|^2) / (4*pi)
%
% lap2d_green()         -- single-threaded (default)
% lap2d_green(N)        -- ask MATLAB to use up to N compute threads
%
% Each rep k = 0..NREP-1 shifts the target X coordinates by k*DT and
% adds the resulting NT-by-NS kernel matrix into an accumulator `acc`.
% The final fingerprint (selected entries, min, max, sum) is taken on
% acc, so every rep contributes to the printed answer -- nothing can
% be hoisted out of the timing loop.

if nargin < 1, nthreads = 1; end

NS = 4000;
NT = 4000;
NREP = 20;
DT = 0.01;

actual_nthreads = nthreads;
try
    maxNumCompThreads(nthreads);
    actual_nthreads = maxNumCompThreads;
catch
    % maxNumCompThreads not available; fall through with requested value
end

[src, targ] = make_inputs(NS, NT);

% Warm up (one full pass through the kernel; result discarded).
acc = compute_kernel(src, targ, 0, DT);  %#ok<NASGU>

% Timed loop: accumulate NREP shifted kernel matrices into acc.
acc = zeros(NT, NS);
t0 = tic;
for k = 0:NREP-1
    acc = acc + compute_kernel(src, targ, k, DT);
end
t = toc(t0);

i_mid = NT/2;
j_mid = NS/2;

vmin = min(acc(:));
vmax = max(acc(:));
csum = sum(acc(:));

fprintf('=== MATLAB lap2d_green ===\n');
fprintf('threads        = %d\n', actual_nthreads);
fprintf('NS=%d  NT=%d  NREP=%d  DT=%.4f\n', NS, NT, NREP, DT);
fprintf('elapsed_total  = %.6f s\n', t);
fprintf('elapsed_per    = %.6f s\n', t/NREP);
fprintf('throughput     = %.3e pts/s\n', NS*NT*NREP/t);
fprintf('acc(1,1)       = %.17e\n', acc(1,1));
fprintf('acc(%d,%d) = %.17e\n', i_mid, j_mid, acc(i_mid, j_mid));
fprintf('acc(NT,NS)     = %.17e\n', acc(NT,NS));
fprintf('min(acc)       = %.17e\n', vmin);
fprintf('max(acc)       = %.17e\n', vmax);
fprintf('sum(acc)       = %.17e\n', csum);
end

function val = compute_kernel(src, targ, k, dt)
% NT-by-NS matrix of -log(|t-s|^2)/(4*pi) with target X shifted by k*dt.
rx = (targ(1,:) + k*dt).' - src(1,:);
ry = targ(2,:).' - src(2,:);
r2 = rx.^2 + ry.^2;
val = -log(r2) / (4*pi);
end

function [src, targ] = make_inputs(NS, NT)
% Deterministic xorshift64 PRNG; sources in [-1,1]^2,
% targets in [-1,1]^2 + (3,0) so |t-s| is bounded away from 0.

state = uint64(2463534242);
src  = zeros(2, NS);
targ = zeros(2, NT);

for j = 1:NS
    [u, state] = xs64(state); src(1,j) = u*2 - 1;
    [u, state] = xs64(state); src(2,j) = u*2 - 1;
end
for j = 1:NT
    [u, state] = xs64(state); targ(1,j) = u*2 - 1 + 3;
    [u, state] = xs64(state); targ(2,j) = u*2 - 1;
end
end

function [u, state] = xs64(state)
% xorshift64 -> uniform double in [0,1)
x = state;
x = bitxor(x, bitshift(x, 13));
x = bitxor(x, bitshift(x, -7));
x = bitxor(x, bitshift(x, 17));
state = x;
u = double(bitshift(x, -11)) * (1.0 / 9007199254740992.0);
end
